README.md

```diff
@@ -796,6 +796,122 @@ The `profunctor` method takes two arguments:
 
 3. `promap` must return a value of the same Profunctor
 
+## Types
+
+In Fantasy Land, types are specified by an object containing a catamorphic
+method `cata`. `cata` has an arity matching the number of constructor functions
+belonging to each type. Each argument to `cata` is either:
+
+1. a value
+2. a function with an arity matching the number of arguments to the constructor
+   function.
+   1. Each function argument to `cata` must return a value of the same type as
+      `cata` itself.
+
+ex.
+```hs
+Identity a = { cata :: (a -> b) -> b }
+identity :: a -> Identity a
+```
+
+Libraries conforming to this specification may provide custom data constructors
+corresponding with the constructor functions. No provided constructors are
+required to share the names of the constructor functions defined herein. For
+example, instead of `Identity`, the constructor could be named `Id`. If custom
+constructors _are_ provided, they must return values which contain a `cata`
+method. This method must exhibit the same behaviour as described above.
+
+ex.
+```js
+var identity = require('fantasy-land/identity');
+
+function Id(x) {
+  this.x = x;
+}
+Id.prototype['fantasy-land/cata'] = function cata(f) {
+  return f(this.x);
+}
+
+// Alternatively
+Id.prototype['fantasy-land/cata'] = function cata(f) {
+  return identity(this.x).cata(f);
+}
+
+(new Id(42)).cata(x => x) === identity(42).cata(x => x);
+```
+
+### Maybe
+
+The `Maybe` type encodes the concept of optionality (Nothing and Just a).
+
+```hs
+Maybe a = { cata :: (b, (a -> b)) -> b }
+nothing :: Maybe a
+just :: a -> Maybe a
+```
+
+A value which conforms to the Maybe specification must provide an `cata` method.
+
+The `cata` method takes two arguments:
+
+    m.cata(x, f)
+
+1. `x` is the default value for the `Nothing` case
+
+    1. No parts of `x` should be checked.
+
+2. `f` must be a function which returns a value
+
+    1. If `f` is not a function, the behaviour of `cata` is unspecified.
+    2. No parts of `f`'s return value should be checked.
+
+### Either
+
+The `Either` type encodes the concept of binary possibility (Left a and Right b).
+
+```hs
+Either a b = { cata :: ((a -> c), (b -> c)) -> c }
+left :: a -> Either a b
+right :: b -> Either a b
+```
+
+A value which conforms to the Either specification must provide an `cata` method.
+
+The `cata` method takes two arguments:
+
+    e.cata(f, g)
+
+1. `f` must be a function which returns a value
+
+    1. If `f` is not a function, the behaviour of `cata` is unspecified.
+    2. No parts of `f`'s return value should be checked.
+
+2. `g` must be a function which returns a value
+
+    1. If `g` is not a function, the behaviour of `cata` is unspecified.
+    2. No parts of `g`'s return value should be checked.
+
+### Tuple
```

Comment:
Should we support tuples with more than two values?

Reply:
`Tuple3 a b c` is isomorphic to `Tuple a (Tuple b c)` or `Tuple a (Tuple b (Tuple c Unit))`. Also, I'd rather not complicate things by adding a specification for every `Tuple{n}`.

Reply:
A general pattern for the definition of tuples _could_ be written and derivations from `Tuple` provided, but that should be a separate PR.

Reply:
What about renaming `Tuple` to `Pair`? That seems more precise since this is a 2-tuple i.e. a pair.

Reply:
To me, the word "tuple" is more general. It covers both 1-tuples (singletons), 2-tuples (pairs) 3-tuples, etc. However this specification is not for tuples in general—it is for 2-tuples in particular. So I think naming it "pair" would be more appropriate.